controller/volume_controller.go

```diff
@@ -1553,6 +1655,260 @@ func getRebuildingReplicaCount(e *longhorn.Engine) int {
 	return rebuilding
 }
 
+type replicaAutoBalanceCount func(*longhorn.Volume, map[string]*longhorn.Replica) (int, map[string][]string, error)
+
+func (vc *VolumeController) replicaAutoBalanceLeastEffort(v *longhorn.Volume,
+	rs map[string]*longhorn.Replica,
+	fnCount replicaAutoBalanceCount) int {
+	log := getLoggerForVolume(vc.logger, v).WithField("replicaAutoBalanceOption", types.ReplicaAutoBalanceLeastEffort)
+
+	var err error
+	defer func() {
+		if err != nil {
+			log.WithError(err).Warn("Skip replica auto-balance")
+		}
+	}()
+
+	setting, err := vc.getAutoBalancedReplicasSetting(v)
+	enabled := []string{
+		string(types.ReplicaAutoBalanceLeastEffort),
+		string(types.ReplicaAutoBalanceBestEffort),
+	}
+	if !util.Contains(enabled, string(setting)) || err != nil {
+		return 0
+	}
+
+	if v.Status.Robustness != types.VolumeRobustnessHealthy {
+		log.Debugf("Cannot auto-balance volume in %s state", v.Status.Robustness)
+		return 0
+	}
+
+	var adjustCount int
+	adjustCount, _, err = fnCount(v, rs)
+	if err != nil {
+		return 0
+	}
+	log.Debugf("Found %v replica candidate for auto-balance", adjustCount)
+	return adjustCount
+}
+
+func (vc *VolumeController) replicaAutoBalanceBestEffort(v *longhorn.Volume,
+	rs map[string]*longhorn.Replica,
+	fnCount replicaAutoBalanceCount) (int, []string) {
+	log := getLoggerForVolume(vc.logger, v).WithField("replicaAutoBalanceOption", types.ReplicaAutoBalanceBestEffort)
+
+	var err error
+	defer func() {
+		if err != nil {
+			log.WithError(err).Warn("Skip replica auto-balance")
+		}
+	}()
+
+	setting, err := vc.getAutoBalancedReplicasSetting(v)
+	if setting != types.ReplicaAutoBalanceBestEffort || err != nil {
+		return 0, nil
+	}
+
+	if v.Status.Robustness != types.VolumeRobustnessHealthy {
+		log.Debugf("Cannot auto-balance volume in %s state", v.Status.Robustness)
+		return 0, nil
+	}
+
+	var unusedCount int
+	unusedCount, idOverlays, err := fnCount(v, rs)
+	if unusedCount == 0 && len(idOverlays) <= 1 {
+		return 0, nil
+	}
+
+	var topCountIDRNames []string
+	var rOverlapTopCount int
+	var rOverlapLowCount int
+	for _, rNames := range idOverlays {
+		rNameCount := len(rNames)
+		if len(rNames) > len(topCountIDRNames) {
+			topCountIDRNames = rNames
+			rOverlapTopCount = rNameCount
+		}
+		if rOverlapLowCount == 0 || rNameCount < rOverlapLowCount {
+			rOverlapLowCount = rNameCount
+		}
+	}
+	adjustCount := rOverlapTopCount - rOverlapLowCount
```

Comment:
Should not the adjustCount equal to `rOverlapTopCount - rOverlapLowCount - 1`?

Reply:
Not sure why include `- 1`? Do you have a specific case where adjustCount is not expected without the `- 1`?

Reply:
Let's take a look at this case again: Assuming there are 2 nodes, nodeA contains 3 replicas while nodeB contains 1 replica. Hence `leastExtraRCount` is 0 while `mostExtraRCount` is 2 if the current logic is applied, then `adjustCount` will be 2. But actually, we only need to adjust 1 replica.

Reply:
You are right, the naming isn't consistent. It's more of a  diff count. Fixed!